### ISSUE TITLE
application hang when call WinDivertClose

### ISSUE THREAD

**Issue by tanda996** [basil00/Divert#298]
![askbug](https://user-images.githubusercontent.com/100284091/155312359-1fa8cc06-1762-4c26-bbd6-9fe70eac40d7.png)
this is my app's stack. My app hang when call WinDivertClose:
if (m_hSocketMon && m_hSocketMon != INVALID_HANDLE_VALUE)
{
	WinDivertClose(m_hSocketMon); // hang here
}

after run above codes, my app still have 3 other handle to close. does Windivert driver call cleanup too soon or somethings?
![image](https://user-images.githubusercontent.com/100284091/155312896-5fbb3196-b821-4e4d-bc6c-fc5584ab4636.png)

**Comment by basil00** [comment]
It seems related to #294.

Eyeballing the code, it is strange that it hangs on `WdfDeleteObject`.  The `read_queue` object should be managed by WDF and not the WinDivert driver.

What version of Windows is this?  Is there a reliable way to reproduce the problem?

**Comment by tanda996** [comment]
Sometime happened on Windows 1703 x86, Windows x64 too but much less.
AFAIK windivert_cleanup() will be called only when all windivert's handles reach to 0 reference. But on my machine it's still have 3 (1 handle hanging, maybe?), this strange.
More info: My app request stop windivert driver (change to stop_pending status) before call WinDivertClose() to close all handles, does this cause to unknown behavior?

**Comment by basil00** [comment]
> AFAIK windivert_cleanup() will be called only when all windivert's handles reach to 0 reference. But on my machine it's still have 3 other handles (1 handle hanging, maybe?), this strange.

`windivert_cleanup()` is called by the WDF framework once the ref count for a given handle drops to zero.  I assume the other handles are independent so should not matter, or else if they are duplicates (`DuplicateHandle`) the ref count should not be zero.

I also checked, and it seems what WinDivert is doing is correct, on [paper at least](https://docs.microsoft.com/en-us/windows-hardware/drivers/wdf/deleting-i-o-queues):

_For example, a driver that provides an [EvtDeviceFileCreate](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/wdfdevice/nc-wdfdevice-evt_wdf_device_file_create) callback function might create an I/O queue to handle I/O requests that are associated with a particular framework file object. In this case the driver's [EvtFileCleanup](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/wdfdevice/nc-wdfdevice-evt_wdf_file_cleanup) callback function must call [WdfIoQueuePurge](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/wdfio/nf-wdfio-wdfioqueuepurge) to purge the queue and then call WdfObjectDelete to delete it._

Basically, when the program calls `WinDivertRecv` there may not be any packet to service the request at that time.  So WinDivert needs to queue the request (waiting for the next packet), which is why it has the `read_queue`.

So the current design appears to be sound at least, and the bug probably is something else.  But I don't think I can make progress on this unless I can replicate the bug.
